instrumentation/src/main.c

```diff
@@ -0,0 +1,101 @@
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdio.h>
+
+#define MAX_LINE_LENGTH 1023
+#define MAX_LINES 50
+
+// TODO change to systemd drop in file paths
+static char *const env_var_file_java = "/etc/splunk/zeroconfig_java.conf";
+static char *const env_var_file_node = "/etc/splunk/zeroconfig_node.conf";
+
+static char *const allowed_env_vars[] = {
+    "OTEL_SERVICE_NAME",
+    "OTEL_EXPORTER_OTLP_ENDPOINT",
+    "OTEL_RESOURCE_ATTRIBUTES",
+    "SPLUNK_PROFILER_ENABLED",
+    "SPLUNK_PROFILER_MEMORY_ENABLED",
+    "SPLUNK_METRICS_ENABLED",
+};
+
+const size_t allowed_env_vars_size = sizeof(allowed_env_vars) / sizeof(*allowed_env_vars);
+
+extern char *program_invocation_short_name;
+
+// The entry point for all executables prior to their execution.
+void __attribute__((constructor)) enter() {
+    char *env_var_file;
+    if (strcmp("java", program_invocation_short_name) == 0) {
+        env_var_file = env_var_file_java;
+    } else if (strcmp("node", program_invocation_short_name) == 0) {
+        env_var_file = env_var_file_node;
+    } else {
+        // we don't want to inject environment variables for this program.
+        return;
+    }
+    const size_t buffer_size = MAX_LINE_LENGTH + 1;
+    char buffer[buffer_size];
+
+    if (MAX_LINES <= 0 || MAX_LINE_LENGTH <= 0) {
+        return;
+    }
+
+    FILE *fp = fopen(env_var_file, "r");
+    if (fp == NULL) {
+        return;
+    }
+
+    int line_count = 0;
+
+    while (fgets(buffer, buffer_size, fp) != NULL) {
+        line_count += 1;
+        if (line_count > MAX_LINES) {
+            break;
+        }
+
+        char *newline = strchr(buffer, '\n');
+        if (newline != NULL) {
+            // terminate the string inside buffer at the newline
+            *newline = '\0';
+        }
+        // if we have read a string without a newline termination, we have reached the end of file
+        // or the string is past our max buffer size and we have an invalid value and we need to abort.
+
+        // Properly formatted input file contains lines of length less than MAX_LINE_LENGTH and ends with a newline.
+        // If newline character is not read, it means that a line is greater that MAX_LINE_LENGTH or the file does not end with a newline.
+        if (newline == NULL) {
+            break;
+        }
+        if (strlen(buffer) == 0) {
```

Comment:
General recommendation: use the "n" versions/alternatives for all functions dealing with `char` buffers in C. Not only `strlen`, but, `memchar` instead of `strchr`, etc.  We can reason that the "n" versions/alternatives aren't needed in some instances, but instead of relying on such reasoning and assumptions, using the "n" versions/alternatives is more straightforward and safer.

Reply:
I can look but I'm bad at C, and I am afraid I don't completely understand what you mean. I'll take a deeper look.

Reply:
OK, I have applied that approach.